csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,271 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+        });
+    // gmem to smem
+    latency += gmem_to_smem_latency * (is_reused ? 1 : n_outer_loop);
+    // smem to register
+    latency += n_outer_loop * n_proj_buffer_uses.at(idx);
+    int dtype_size = (int)dataTypeSize(input_buffer->getDataType().value());
+    buffer_latency[idx] = latency * dtype_size;
+  }
+
+  // reorder projectable_buffer_inputs based on the latency from smallest to
+  // largest
+  std::vector<int> idxs(n_buffer_inputs);
+  std::iota(idxs.begin(), idxs.end(), 0);
+  std::stable_sort(idxs.begin(), idxs.end(), [&buffer_latency](int i, int j) {
+    return buffer_latency[i] < buffer_latency[j];
+  });
+  std::vector<TensorView*> sorted_candidate_tvs;
+  sorted_candidate_tvs.reserve(n_buffer_inputs);
+  for (auto idx : idxs) {
+    sorted_candidate_tvs.emplace_back(projectable_buffer_inputs.at(idx));
+  }
+  return sorted_candidate_tvs;
+}
+
+// Sort persistent buffers based on the number of uses from smallest to largest.
+// Used when buffers are not projected to inputs.
+// TODO: Currently, this function is unused as all existing cases (LayerNorm bwd
+// and RMS norm bwd) involve projecting buffers to inputs. Consider this
+// function a placeholder for scenarios where buffers are not projected to
+// inputs. Refer to the test CombinedSchedulerInnerOuterNoOuterBroadcastTv for a
+// concrete example of such cases. Add NoOuterBroadcast cases to our benchmarks
+// and test the performance of this function.
+std::vector<TensorView*> sortPersistentBuffers(
+    std::vector<TensorView*> persistent_buffers) {
+  // Directly sort the persistent_buffers vector based on the number of uses.
+  std::stable_sort(
+      persistent_buffers.begin(),
+      persistent_buffers.end(),
+      [](TensorView* a, TensorView* b) {
+        return a->uses().size() < b->uses().size();
+      });
+  return persistent_buffers;
+}
+
+// Decide where to store persistent buffers.
+// By default, they reside in registers.
+// If register space runs low but there's ample shared memory,
+// move one or more buffers to shared memory. Since shared memory
+// is slower than registers, we should move the buffers that are
+// less frequently accessed to shared memory.
+struct PersistentBufferStorageParams {
+  // representing buffers that are stored in shared memory, other buffers are
+  // stored in registers.
+  std::vector<TensorView*> smem_persistent_buffers;
+
+  // Total number of bytes occupied by all persistent buffers stored in shared
+  // memory.
+  int64_t smem_buffer_size = -1;
+
+  // Total number of bytes occupied by all persistent buffers stored in
+  // registers.
+  int64_t regs_buffer_size = -1;
+
+  // Additional shared memory usage per block that is not associated with
+  // persistent buffers. This includes memory for driver overhead and workspace
+  // for reductions.
+  int64_t smem_overhead = -1;
+
+  // Flag indicating whether there are sufficient registers and shared memory
```

Comment:
What happens when this flag is false? Is the scheduler going to be rejected?

Reply:
yes, false means we don't have sufficient registers and shared memory.